runtime/compilers/rillv1/parser.go

```diff
@@ -0,0 +1,603 @@
+	// The logic here is slightly tricky because the relationship between files and resources can vary:
+	//
+	// - Case 1: one file created one resource
+	// - Case 2: one file created multiple resources
+	// - Case 3: multiple files contributed to one resource (for example, "model.sql" and "model.yaml")
+	//
+	// The high-level approach is: We'll delete all existing resources *related* to the changed paths and (re)parse them.
+	// Then at the end, we build a diff that treats any resource that was both "deleted" and "added" as an "update".
+	// (Renames are not supported in the parser. It needs to be handled by the caller, since parser state alone is insufficient to detect it.)
+
+	// Phase 1: Clear existing state related to the paths.
+	// Identify all paths directly passed and paths indirectly related through resourcesForPath and Resource.Paths.
+	// And delete all resources and parse errors related to those paths.
+	var parsePaths []string            // Paths we should pass to parsePaths
+	var deletedResources []*Resource   // Resources deleted in Phase 1 (some may be added back in Phase 2)
+	checkPaths := slices.Clone(paths)  // Paths we should visit in the loop
+	seenPaths := make(map[string]bool) // Paths already visited by the loop
+	for i := 0; i < len(checkPaths); i++ {
+		// Don't check the same path twice
+		path := normalizePath(checkPaths[i])
+		if seenPaths[path] {
+			continue
+		}
+		seenPaths[path] = true
+
+		// Skip files that aren't SQL or YAML
+		isSQL := strings.HasSuffix(path, ".sql")
+		isYAML := strings.HasSuffix(path, ".yaml") || strings.HasSuffix(path, ".yml")
+		if !isSQL && !isYAML {
+			continue
+		}
+
+		// If a file exists at path, add it to the parse list
+		_, err := p.Repo.Stat(ctx, p.InstanceID, path)
+		if err == nil {
+			parsePaths = append(parsePaths, path)
+		} else if !os.IsNotExist(err) {
+			return nil, fmt.Errorf("unexpected file stat error: %w", err)
+		}
+
+		// Check if path is rill.yaml and clear it (so we can re-parse it)
+		if path == "/rill.yaml" || path == "/rill.yml" {
+			p.RillYAML = nil
+		}
+
+		// Since .sql and .yaml files provide context for each other, if one was modified, we need to reparse both.
+		// For cases where a file was modified or deleted, the transitive check through resourcesForPath will already take of that.
+		// But this ensures the check also happens for cases where a companion file was added.
+		stem := pathStem(path)
+		if isSQL {
+			checkPaths = append(checkPaths, stem+".yaml", stem+".yml")
+		} else if isYAML {
+			checkPaths = append(checkPaths, stem+".sql")
+		}
+
+		// Remove all resources derived from this path, and add any related paths to the check list
+		for _, resource := range p.resourcesForPath[path] {
+			// Multiple entries in resourcesForPath may point to the same resource.
+			// By adding resource.Paths to checkPaths, the outer loop will eventually clear those (maybe it already has).
+			checkPaths = append(checkPaths, resource.Paths...)
```

Comment:
Should we do the deduplication here? Embedded sources seem to explode checkPaths quite a bit. The test with 2 models referencing once source leads to about 14 entries in this when one of the model is deleted.

Reply:
Good observation. I tried investigating some simple deduplication checks / other logic changes, but wasn't able to come up with something that benchmarked better. For small list sizes, it seems appending to checkPaths is about the same cost as doing extra deduplication checks.

Given the small number of files that will usually be reparsed, I think it's okay to keep it slightly unoptimized (the seenPaths will still guard against duplicating any expensive ops, like checking file stat), and it's nice to keep the logic simpler